### PR TITLE
NET-5530 Support response header modifiers on http-route config entry

### PR DESCRIPTION
**Pull request by nathancoleman**
### Description
Allow an `http-route` config entry to include header modifications for the response in addition to the request header modifications that we already support.

The shape of this addition is consistent with the request filters that we support today but uses a separate field w/ the `Response` prefix since it's nonsensical to have all of the same filters supported that we do on the request:
```hcl
route := api.HTTPRouteConfigEntry{
	Kind: structs.HTTPRoute,
	Name: "http-route",
	Rules: []structs.HTTPRouteRule{{
		Filters: structs.HTTPFilters{
			Headers: []structs.HTTPHeaderFilter{
				{
					Add:    map[string]string{"add me on the request": "hello world"},
					Set:    map[string]string{"set me on the request": "hello world"},
					Remove: []string{"remove me on the request"},
				},
			},
		},
		// =========== THIS IS NEW ===========
		ResponseFilters: structs.HTTPResponseFilters{
			Headers: []structs.HTTPHeaderFilter{
				{
					Add:    map[string]string{"add me on the response": "hello world"},
					Set:    map[string]string{"set me on the response": "hello world"},
					Remove: []string{"remove me on the response"},
				},
			},
		},
		Services: []structs.HTTPService{{ Name: "foo" }},
	}},
}
```

### Testing & Reproduction steps
- [x] Robot tests pass, including updated unit tests
- [x] Can create a `http-route` config entry in Consul with the new request filter property

### PR Checklist

* [x] updated test coverage
* [ ] external facing docs updated
* [x] appropriate backport labels added
* [x] not a security concern

### CODE REVIEW

**Review by andrewstucki — COMMENTED** [review]
This generally looks pretty straightforward. A couple of things needed:

1. A changelog entry, and
2. Some tests. Could we get at minimum a fixtures-based test through leveraging proxycfg? Or adding some sort of testing onto an existing integration test actually leveraging the whole stack?

After those, I'm good on approving.

**Comment by nathancoleman** [comment]
@andrewstucki I decided on expanding one of our existing integration tests to cover response headers

File: agent/consul/discoverychain/gateway_httproute.go [hunk 1]
```diff
@@ -90,16 +91,29 @@ func httpRouteToDiscoveryChain(route structs.HTTPRouteConfigEntry) (*structs.Ser
...
 				servicePrefixRewrite = prefixRewrite
 			}
-			serviceModifier := httpRouteFiltersToServiceRouteHeaderModifier(service.Filters.Headers)
-			modifier.Add = mergeMaps(modifier.Add, serviceModifier.Add)
-			modifier.Set = mergeMaps(modifier.Set, serviceModifier.Set)
-			modifier.Remove = append(modifier.Remove, serviceModifier.Remove...)
+
+			// Merge service request header modifier(s) onto route rule modifiers
+			// Note: Removals for the same header may exist on the rule + the service and
+			//   will result in idempotent duplicate values in the modifier w/ service coming last
```
Comment (nathancoleman) [inline agent/consul/discoverychain/gateway_httproute.go:97]: Calling this out as pre-existing behavior. I considered doing a merge of the `Remove` lists but did not want to introduce that as a side effect of this PR

File: agent/consul/discoverychain/gateway_test.go
Comment (nathancoleman) [inline agent/consul/discoverychain/gateway_test.go:1]: Changes in this test:
- add coverage for request filter merging where there was none before
- verify that breaking the modifiers across multiple filter entries works as expected
- verify that service-level modifiers take precedence over rule-level modifiers
- demonstrate the duplicate remove modifier behavior described in the docstrings added above in `agent/consul/discoverychain/gateway_httproute.go`

File: agent/consul/discoverychain/gateway_test.go [hunk 2]
```diff
@@ -557,8 +619,40 @@ func TestGatewayChainSynthesizer_Synthesize(t *testing.T) {
 									Partition: "default",
 									Namespace: "default",
 									RequestHeaders: &structs.HTTPHeaderModifiers{
-										Add: make(map[string]string),
-										Set: make(map[string]string),
+										Add: map[string]string{
+											"add me to the rule request":             "present",
+											"add me to the service request":          "present",
+											"add me to the rule and service request": "service",
```
Comment (nathancoleman) [inline agent/consul/discoverychain/gateway_test.go:625]: we expect `service` as the value here, indicating that the filter modifier on the service took precedence over the same modifier on the rule which specifies `rule` as the value.

this pattern applies to the other modifiers below as well.

**Review by andrewstucki — APPROVED** [review]

**Comment by nathancoleman** [comment]
Have decided retroactively not to backport. The backport was already merged but is being reverted in #18770

**Pull request merged**